### PR TITLE
Add the in-memory metrics exporter.

### PR DESCRIPTION
**Pull request by yxue**
Fixes # (issue) [1405](https://github.com/open-telemetry/opentelemetry-cpp/issues/1405)

## Changes

Please provide a brief description of the changes here.

For significant contributions please make sure you have completed the following items:

* [x] `CHANGELOG.md` updated for non-trivial changes
* [x] Unit tests have been added
* [x] Changes in public API reviewed

### CODE REVIEW

**Comment by owent** [comment]
Could you please alse add cmake targets in `exporters/memory/CMakeLists.txt` and `cmake/opentelemetry-cpp-config.cmake.in` .

**Comment by yxue** [comment]
> Could you please alse add cmake targets in `exporters/memory/CMakeLists.txt` and `cmake/opentelemetry-cpp-config.cmake.in` .

Sure, I can add that. I feel supporting/maintaining two build systems is a little bit inconvenient. Do we consider using some tools like, https://github.com/google/bazel-to-cmake, so that CMakeLists.txt can be generated from the BUILD and WORKSPACE, or using cmake and bazel (https://github.com/bazelbuild/rules_foreign_cc) supports importing cmake projects.

**Comment by owent** [comment]
> > Could you please alse add cmake targets in `exporters/memory/CMakeLists.txt` and `cmake/opentelemetry-cpp-config.cmake.in` .
> 
> Sure, I can add that. I feel supporting/maintaining two build systems is a little bit inconvenient. Do we consider using some tools like, https://github.com/google/bazel-to-cmake, so that CMakeLists.txt can be generated from the BUILD and WORKSPACE, or using cmake and bazel (https://github.com/bazelbuild/rules_foreign_cc) supports importing cmake projects. 

Good point, I think it require more discussion. In most case, I think cmake users do not want to install another bazel toolset, and bazel users do not want install cmake toolset either.
cmake has many modules and tools to work with legacy build toolchains(pkg-config, makefile, autoconf, android ndk, feature dection by test running or compiling some c/c++ codes, environments and etc.).I'm not family with bazel and I'm not sure if it's easy to let bazel and cmake work together with these modules.

BTW: We had a discussion before and try to find a way to use bazel to build both with legacy gcc 4.8 and the latest compiler here https://github.com/open-telemetry/opentelemetry-cpp/pull/666 , do you has any suggestion about it?

**Comment by lalitb** [comment]
>  I feel supporting/maintaining two build systems is a little bit inconvenient

Yes, I do agree it's inconvenient to maintain both these build systems. The decision to support both was made much before any of the existing contributors joined the team. The SIG work was initiated by the folks from Google and Microsoft, and it was defacto choice to support the build systems used by these companies :). Having said that, most of the existing C++ projects use either(or both) of these build systems, and so it makes sense to keep supporting both the build systems.

**Comment by ThomsonTan** [comment]
The 2 build systems issue were discussed in the past as @lalitb and @owent mentioned. As there are users for both of them, we will try our best to keep both up to date and consistent. For the https://github.com/google/bazel-to-cmake, it has not been updated for a few years, not sure it could work, together with that we've already have full set of cmake files which is not expected to be overwriten.

**Comment by yxue** [comment]
> 



> Thanks for the PR @yxue. This is your first PR here, and it is nicely done :)
> 
> Just thinking whether the InMemoryExporter should have any dependency on the protobuf. For Logs and Traces, the memory/ostream exporters are not dependent on any external library. Is it possible to achieve this for Metrics too - Eg by letting CircularBuffer contain the list of [MetricData ](https://github.com/open-telemetry/opentelemetry-cpp/blob/8691a2d582c2e0c811fbf383b572e1ff882b5b37/sdk/include/opentelemetry/sdk/metrics/data/metric_data.h#L30) or [ResourceMetrics ](https://github.com/open-telemetry/opentelemetry-cpp/blob/8691a2d582c2e0c811fbf383b572e1ff882b5b37/sdk/include/opentelemetry/sdk/metrics/export/metric_producer.h#L27) ?
> 
> We can discuss further if it is something not feasible/or difficult to achieve with the current SDK implementation.

Thanks @lalitb! I was thinking about using the CircularBuffer with the  ResourceMetrics but then we need the special process for the pointers inside the structure.  

Having the MetricData with CircularBuffer might work but some info is lost. 

Another thing is about comparing these structure objects. With protobuf message, the objects can be easily compared. Since we don't define the == for MetricsData and ResourceMetrics, comparing them in the unit test is not trivial.

**Comment by yxue** [comment]
Thanks @lalitb @owent and @ThomsonTan for the info about the building system. 

I am not sure about fixing the gcc 4.8 and grpc compatibility. I do see the building systems are right now used by multiple teams/companies. I also saw there are already some divergence problems between these two systems, e.g., https://github.com/open-telemetry/opentelemetry-cpp/issues/1428. I think we might need some tools to reconcile that if the community is going to keep two building systems.

**Comment by lalitb** [comment]
> Thanks @lalitb! I was thinking about using the CircularBuffer with the ResourceMetrics but then we need the special process for the pointers inside the structure

Yes, I thought so too. Should be fine to keep the dependency unless someone has any concern :)
btw, I didn't realize this PR also has changes for bazel build for metrics proto files and raised a PR for that in #1489. Sorry about that, as you would have to resolve the conflict here :( Would ensure not to merge any further conflicting PRs till this is merged.

**Comment by yxue** [comment]
> > Thanks @lalitb! I was thinking about using the CircularBuffer with the ResourceMetrics but then we need the special process for the pointers inside the structure
> 
> Yes, I thought so too. Should be fine to keep the dependency unless someone has any concern :) btw, I didn't realize this PR also has changes for bazel build for metrics proto files and raised a PR for that in #1489. Sorry about that, as you would have to resolve the conflict here :( Would ensure not to merge any further conflicting PRs till this is merged.

Sure, no worries. The PR was updated. PTAL, thanks!

File: exporters/memory/test/in_memory_metric_exporter_test.cc [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+// SPDX-License-Identifier: Apache-2.0
+#ifndef ENABLE_METRICS_PREVIEW
+
+#  include "opentelemetry/exporters/memory/in_memory_metric_exporter.h"
+#  include "google/protobuf/text_format.h"
+#  include "opentelemetry/sdk/metrics/data/metric_data.h"
+
+#  include <gmock/gmock.h>
+#  include <gtest/gtest.h>
+#  include <iostream>
```
Comment (lalitb) [inline exporters/memory/test/in_memory_metric_exporter_test.cc:11]: do we need iostream?

**Review by lalitb — APPROVED** [review]
LGTM.

File: exporters/memory/src/in_memory_metric_exporter.cc [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  std::unique_ptr<proto::metrics::v1::ResourceMetrics> resource_metrics(
+      new proto::metrics::v1::ResourceMetrics);
+  otlp::OtlpMetricsUtils::PopulateResourceMetrics(data, resource_metrics.get());
+  data_->Add(std::move(resource_metrics));
+  return sdk::common::ExportResult::kSuccess;
+}
+
+bool InMemoryMetricExporter::ForceFlush(std::chrono::microseconds timeout) noexcept
+{
+  const std::lock_guard<opentelemetry::common::SpinLockMutex> locked(lock_);
```
[the remark this answers is not in the document]
Comment (yxue) [reply exporters/memory/src/in_memory_metric_exporter.cc]: done

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_data.h [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace exporter
+{
+namespace memory
+{
+/**
+ * A wrapper class holding in memory exporter data
+ */
+template <typename T>
```
[the remark this answers is not in the document]
Comment (yxue) [reply exporters/memory/include/opentelemetry/exporters/memory/in_memory_data.h:18]: sure, https://github.com/open-telemetry/opentelemetry-cpp/issues/1493 created

**Comment by owent** [comment]
Could you please revert changes in `/exporters/otlp` ?It's duplicated with #1487 , I belive this PR is just for memory metrics exporter.

**Comment by yxue** [comment]
> Could you please revert changes in `/exporters/otlp` ?It's duplicated with #1487 , I belive this PR is just for memory metrics exporter.

The change under exporters/otlp is for fixing the missing part of conversion between protobuf and structure. Remove that will fail the unit test.

**Comment by owent** [comment]
> > Could you please revert changes in `/exporters/otlp` ?It's duplicated with #1487 , I belive this PR is just for memory metrics exporter.
> 
> The change under exporters/otlp is for fixing the missing part of conversion between protobuf and structure. Remove that will fail the unit test.

OK. we can solve this conflict later.

**Review by owent — CHANGES_REQUESTED** [review]
LGTM after `in_memory_metric_exporter` is also added into `cmake/opentelemetry-cpp-config.cmake.in` ?

**Comment by lalitb** [comment]
> LGTM after `in_memory_metric_exporter` is also added into `cmake/opentelemetry-cpp-config.cmake.in`.

Good point. @yxue this file contains the rules to enable opentelemetry-cpp included in other projects cmake files using `find_package(opentelemetry-cpp)`. All the targets/libraries created by opentelemetry-cpp should be included in this file.

**Comment by yxue** [comment]
> > LGTM after `in_memory_metric_exporter` is also added into `cmake/opentelemetry-cpp-config.cmake.in`.
> 
> Good point. @yxue this file contains the rules to enable opentelemetry-cpp included in other projects cmake files using `find_package(opentelemetry-cpp)`. All the targets/libraries created by opentelemetry-cpp should be included in this file.

Updated, thanks!

**Review by owent — APPROVED** [review]

File: exporters/memory/CMakeLists.txt [hunk 4]
```diff
@@ -21,19 +21,56 @@ install(
...
+  add_library(
+    opentelemetry_exporter_in_memory_metric
+    src/in_memory_metric_exporter.cc ../otlp/src/otlp_metric_utils.cc
+    ../otlp/src/otlp_populate_attribute_utils.cc)
+
+  target_include_directories(
+    opentelemetry_exporter_in_memory_metric
+    PUBLIC "$<BUILD_INTERFACE:${CMAKE_CURRENT_LIST_DIR}/include>"
+           "$<INSTALL_INTERFACE:include>"
+           "$<BUILD_INTERFACE:${CMAKE_CURRENT_LIST_DIR}/../otlp/include>")
```
Comment (owent) [inline exporters/memory/CMakeLists.txt:41]: `target_link_libraries(opentelemetry_exporter_in_memory_metric PUBLIC opentelemetry_otlp_recordable)` would be a better alternative here.But when we building without `WITH_OTLP`, `opentelemetry_otlp_recordable` will be unavailable.
Maybe this module should also be disabled when we build without `WITH_OTLP` now?

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_metric_exporter.h [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  bool ForceFlush(
+      std::chrono::microseconds timeout = (std::chrono::microseconds::max)()) noexcept override;
+
+  bool Shutdown(
+      std::chrono::microseconds timeout = (std::chrono::microseconds::max)()) noexcept override;
+
+  /**
+   * @return Returns a shared pointer to this exporters InMemoryMetricData.
+   */
+  inline InMemoryMetricData &GetData() noexcept { return *data_; }
```
Comment (marcalff) [inline exporters/memory/include/opentelemetry/exporters/memory/in_memory_metric_exporter.h:55]: The comment is outdated, this does not return a shared pointer.

**Review by marcalff — COMMENTED** [review]
LGTM

**Comment by ThomsonTan** [comment]
@yxue could you please resolve the merge conflicts?

**Review by marcalff — CHANGES_REQUESTED** [review]
As discussed in C++ SIG meeting,
the in memory storage structure should not depend on protobuf:

```
  using InMemoryMetricData =
      opentelemetry::exporter::memory::InMemoryData<proto::metrics::v1::ResourceMetrics>;

```
Some refactoring needed to address this part.

**Pull request closed without merging**